STORM-2369 [storm-redis] Use binary type for State management (1.x)

HeartSaVioR opened the pull request:
* introduce new command interface and relevant containers
  * new command interface will contain both binary and string commands once we address STORM-2304
* change RedisKeyValue to use byte[] for State key and value management instead of String
  * get rid of Base64 encode/decode on RedisEncoder, and also SafeEncoder.encode() in Jedis internal
* implement some utils for handling Map with byte[]: since byte[] is bad for key type of Map
* users can select either Redis single server or Redis Cluster by passing JedisPoolConfig or JedisClusterConfig

PR for master: #2172

HeartSaVioR commented:
Tested storm-starter's StatefulTopology with Redis Cluster nodes manually.

HeartSaVioR commented:
Might need to provide migration tool to migrate Base64 encoded states to be conformed to current implementation.

HeartSaVioR commented:
@arunmahadevan 
Please take a look when you have some time. Thanks in advance!

arunmahadevan commented:
@HeartSaVioR what is the motivation for using binary instead of string type ? Is this mainly for performance ? I will review this sometime next week.

HeartSaVioR commented:
@arunmahadevan 
We could save 
- cost of serde. on KV pair
- another serde. for String to byte[] while sending/receiving from/to Redis
- traffic/space of Redis since base64 converts KV to much longer

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -152,70 +170,72 @@ public V get(K key, V defaultValue) {
             }
             if (!currentPending.isEmpty()) {
-                commands.hmset(prepareNamespace, currentPending);
+                commands.hmset(prepareNamespace, ByteArrayUtil.Maps.newHashMapUnwrappingKey(currentPending));
```
maybe you can avoid the wrapping/unwrapping and extra copying if you use a TreeMap with a comparator that compares the byte array. (e.g. `ConcurrentSkipListMap<>(com.google.common.primitives.UnsignedBytes.lexicographicalComparator())` since we want concurrency).

HeartSaVioR replied:
Great suggestion. I thought it is bad but didn't know about alternatives. I'll try to replace that one.

HeartSaVioR commented on external/storm-redis/src/main/java/org/apache/storm/redis/utils/ByteArrayUtil.java:
```diff
@@ -0,0 +1,61 @@
+            return destination;
+        }
+
+        public static Map<ByteArrayWrapper, byte[]> newConcurrentHashMapWrappingKey(Map<byte[], byte[]> source) {
```
Yes I'll remove that.

HeartSaVioR commented on external/storm-redis/src/main/java/org/apache/storm/redis/common/commands/RedisCommands.java:
```diff
@@ -0,0 +1,56 @@
+    ScanResult<Map.Entry<byte[], byte[]>> hscan(byte[] key, byte[] cursor, ScanParams params);
+
+    // string, common
+    boolean exists(String key);
```
Good suggestion. I'll group them per command.

HeartSaVioR commented:
Agreed to leave a note regarding that, and I'm OK to provide a tool to convert current state since state KVs would be valuable for users. I'll try to handle that.

HeartSaVioR commented:
@arunmahadevan 
Sorry got really late for addressing review comments. 

Addressed comments, and created migration tool `Base64ToBinaryStateMigrationUtil`.
(Succeed to read State with this patch from migrated State.)

I'll describe how to use to the doc soon.

HeartSaVioR commented:
@arunmahadevan 
Also updated the doc to explain how to use migration tool. Commits squashed. 
Please review again. Thanks in advance.

HeartSaVioR commented:
Rebased and updated state iterator implementation.
Also extracted base implementation of binary based state iterator into storm-core which could be used also from HBase state.

@arunmahadevan Please take a look when you have time. Thanks in advance.

arunmahadevan commented on storm-core/src/jvm/org/apache/storm/state/DefaultStateEncoder.java:
```diff
@@ -15,27 +15,23 @@
+ * Helper class for encoding/decoding key values.
  */
-public class RedisEncoder<K, V> {
+public class DefaultStateEncoder<K, V> {
```
Maybe this can the default implementation of a StateEncoder interface?

HeartSaVioR replied:
We're already exposing Serializer for key and value, so not sure we want to have another interface and implementations here. Do we want to have encoder for storage which doesn't support binary?

arunmahadevan replied:
Was thinking the `byte[]` is specific to Redis. Do we expect to encode in similar format for other implementations as well? Maybe we can refactor it later if we need different formats for other implementations.

HeartSaVioR replied:
HBase uses `byte[]` as raw type so we could use the same encoder to HBase State, too.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -64,75 +75,83 @@ public RedisKeyValueState(String namespace, JedisPoolConfig poolConfig) {
             } else {
                 LOG.debug("No previously prepared commits.");
-                pendingCommit = Collections.emptyMap();
+                pendingCommit = createPendingCommitMap();
```
nit: can be `Collections.emptyNavigableMap()` (java8) or a static final variable initialized with an empty TreeMap

HeartSaVioR replied:
Good point. Will address. Btw I'll try to avoid JDK 8 thing right now unless it is only applied to the master branch.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -187,70 +206,73 @@ public void prepareCommit(long txid) {
             commands.hmset(txidNamespace, txIds);
             commands.del(prepareNamespace);
-            pendingCommit = Collections.emptyMap();
+            pendingCommit = createPendingCommitMap();
```
nit: can be `Collections.emptyNavigableMap()` (java8) or a static final variable initialized with an empty TreeMap

HeartSaVioR replied:
Will address.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -266,10 +288,10 @@ public void rollback() {
             }
-            pendingCommit = Collections.emptyMap();
-            pendingPrepare = new ConcurrentHashMap<>();
+            pendingCommit = createPendingCommitMap();
```
nit: can be `Collections.emptyNavigableMap()` (java8) or a static final variable initialized with an empty TreeMap

HeartSaVioR replied:
Will address.

arunmahadevan commented on storm-core/src/jvm/org/apache/storm/state/BaseBinaryStateIterator.java:
```diff
@@ -0,0 +1,162 @@
+   *
+   * @return Iterator of loaded state KVs
+   */
+  protected abstract Iterator<Map.Entry<byte[],byte[]>> loadChunkFromStateStorage();
```
Here it assumes the state implementations store in binary format? Can't it return `Iterator<Map.Entry<K, V>>` instead and not assume any specific implementation and let sub-classes do the decoding ?

HeartSaVioR replied:
The type of pendingPrepare and pendingCommit, and storage should be same, so the type is bound to also constructor parameters. So in order to make this generic, we need to have two additional types for storage raw type. I'll see how it will be changed when it is generalized.

arunmahadevan commented on storm-core/src/jvm/org/apache/storm/state/BaseBinaryStateIterator.java:
```diff
@@ -0,0 +1,162 @@
+   * @param key byte array encoded key
+   * @return Decoded value of key
+   */
+  protected abstract K decodeKey(byte[] key);
```
Can we let the subclass deal with the specifics of decoding than assuming data is encoded as `byte[]`?

HeartSaVioR replied:
Same above.

arunmahadevan commented on storm-core/src/jvm/org/apache/storm/state/BaseBinaryStateIterator.java:
```diff
@@ -0,0 +1,162 @@
+   * @param value byte array encoded value
+   * @return Decoded value of value
+   */
+  protected abstract V decodeValue(byte[] value);
```
Can we let the subclass deal with the specifics of decoding than assuming data is encoded as `byte[]`?

HeartSaVioR replied:
Same above.

HeartSaVioR commented on docs/State-checkpointing.md:
````diff
@@ -70,6 +70,46 @@ json config with the following properties.
+ }
+```
+
+NOTE: If you used Redis state with Storm version 1.1.0 or earlier, you would need to also migrate your state since the representation of state has changed  
````
I think it will throw exception because state will not be converted properly, but there's no code to determine State version for now. Adding state version to the checkpoint spout state might be the one way to do this. I'll look into more about how we can handle this.

HeartSaVioR replied:
I'm trying to address this via adding `version` field to CheckpointState class, but stuck on migrating old checkpoint spout state to newer. Kyro doesn't support forward compatibility, and even seemed not be able to read with copied old class. I have no idea how to do it right now.

HeartSaVioR commented on docs/State-checkpointing.md:
````diff
@@ -70,6 +70,46 @@ json config with the following properties.
+    --password <arg>    Redis password (default: no password)
+```
+
+You can provide multiple `namespace` options to migrate multiple namespaces at once. Other options are not mandatory.
````
Yes that would be nice to add it. Will add.

HeartSaVioR commented:
@arunmahadevan Thanks for the further review comments. I addressed your comments without versioning state.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -316,10 +340,18 @@ private Long lastPreparedTxid() {
+        return new ConcurrentSkipListMap<>(UnsignedBytes.lexicographicalComparator());
+    }
+
+    private NavigableMap<byte[], byte[]> createPendingCommitMap() {
```
Seems this is not used. If so can be removed.

HeartSaVioR commented:
@arunmahadevan Thanks I've addressed latest review comment and also squashed.

HeartSaVioR commented:
@arunmahadevan Ah, please review #2172 as well when you revisit this. Thanks in advance!

arunmahadevan commented:
+1
